ext/ddtrace.c

```diff
@@ -621,6 +621,139 @@ static PHP_GSHUTDOWN_FUNCTION(ddtrace) {
+
+    Z_TRY_ADDREF(event->property_name);
+    zend_hash_add(array, name, &event->property_name);
+    Z_TRY_ADDREF(event->property_timestamp);
+    zend_hash_add(array, timestamp, &event->property_timestamp);
+
+    zend_string_release(name);
+    zend_string_release(timestamp);
+
+    if (Z_TYPE(event->property_attributes) != IS_NULL && Z_ARRVAL(event->property_attributes)->nNumOfElements > 0) {
+        zend_string *attributes = zend_string_init("attributes", sizeof("attributes") - 1, 0);
+        Z_TRY_ADDREF(event->property_attributes);
+        zend_hash_add(array, attributes, &event->property_attributes);
+        zend_string_release(attributes);
+    }
+
+    RETURN_ARR(array);
+}
+
+PHP_METHOD(DDTrace_SpanEvent, __construct) {
+    UNUSED(return_value);
+
+    zend_string *name;
+    zval *attributes = NULL;
+    zend_long timestamp = 0;
+
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+        Z_PARAM_STR(name)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_ARRAY_EX(attributes, 1, 0)
+        Z_PARAM_LONG(timestamp)
+    ZEND_PARSE_PARAMETERS_END();
+
+    ddtrace_span_event *event = (ddtrace_span_event*)Z_OBJ_P(ZEND_THIS);
+
+    ZVAL_STR_COPY(&event->property_name, name);
+
+    // Use the provided timestamp or the current time in nanoseconds
+    if (timestamp == 0) {
+        struct timespec ts;
+        timespec_get(&ts, TIME_UTC);
+        ZVAL_LONG(&event->property_timestamp, ts.tv_sec * ZEND_NANO_IN_SEC + ts.tv_nsec);
+    } else {
+        ZVAL_LONG(&event->property_timestamp, timestamp);
+    }
+
+    if (attributes && Z_TYPE_P(attributes) == IS_ARRAY) {
+        ZVAL_DUP(&event->property_attributes, attributes);
+    } else {
+        array_init(&event->property_attributes);
+    }
+}
+
+/* DDTrace\ExceptionSpanEvent */
+zend_class_entry *ddtrace_ce_exception_span_event;
+
+PHP_METHOD(DDTrace_ExceptionSpanEvent, __construct)
+{
+    UNUSED(return_value);
+
+    zval *exception;
+    zval *attributes = NULL;
+
+    ZEND_PARSE_PARAMETERS_START(1, 2)
+        Z_PARAM_OBJECT_OF_CLASS(exception, zend_ce_throwable)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_ARRAY_EX(attributes, 1, 0)
+    ZEND_PARSE_PARAMETERS_END();
+
+    // Prepare parameters for parent constructor
+    zval parent_construct_params[2];
+
+    // First parameter: name (fixed string "exception")
+    zend_string *name = zend_string_init("exception", sizeof("exception") - 1, 0);
+    ZVAL_STR(&parent_construct_params[0], name);
+
+    // Second parameter: attributes
+    if (attributes && Z_TYPE_P(attributes) == IS_ARRAY) {
+        ZVAL_DUP(&parent_construct_params[1], attributes);
+    } else {
+        array_init(&parent_construct_params[1]);
+    }
+
+    // Call the parent constructor with the prepared parameters
+    zval parent_construct_retval;
+    zend_fcall_info fci;
+    zend_fcall_info_cache fci_cache;
+
+    zend_class_entry *parent_ce = Z_OBJCE_P(ZEND_THIS)->parent;
+    zend_function *parent_constructor = zend_hash_str_find_ptr(
```

Comment:
Why so complicated and not simply extract a static function `void dd_span_event_construct(zend_string *name, zend_array *attributes, zend_long timestamp)` which handles setting the name and the attributes.

Also the code here doesn't assign the exception property at all, right?